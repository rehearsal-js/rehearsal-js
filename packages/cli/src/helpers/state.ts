import { resolve } from 'path';
import { existsSync, writeJSONSync, readJSONSync, readFileSync, mkdirSync } from 'fs-extra';
import execa from 'execa';

// The reaosn to have packageMap is getting all files in a package faster
// than loop through everyhing in the files
export type Store = {
  name: string | null;
  packageMap: PackageMap;
  files: FileStateMap;
};

// represent single file state
type FileState = {
  origin: string;
  current: string | null;
  package: string;
  errorCount: number;
};

// use file path as key
export type FileStateMap = {
  [key: string]: FileState;
};

// help to get all files from a package instead of looping all files
type PackageMap = {
  [key: string]: string[];
};

type PackageMigrateProgress = {
  migratedFileCount: number;
  totalFileCount: number;
  isCompleted: boolean;
};

const DEFAULT_REHEARSAL_PATH = resolve(process.cwd(), '.rehearsal');
const DEFAULT_CONFIG_PATH = resolve(DEFAULT_REHEARSAL_PATH, 'migrate-state.json');
const REHEARSAL_TODO_REGEX = /@rehearsal TODO/g;

export class State {
  private configPath: string;
  private store: Store;
  constructor(
    name: string | null,
    packages: string[] = [],
    configPath: string = DEFAULT_CONFIG_PATH
  ) {
    this.configPath = configPath;
    let store;
    if (this.isStateExists()) {
      // every time loading a previous state, need to check if files on disk matches the state
      store = this.getVerifiedStore();
    } else {
      const initPackageMap: PackageMap = {};
      const packageMap: PackageMap = packages.reduce((map, current) => {
        map[current] = [];
        return map;
      }, initPackageMap);
      store = { name: name ? name : '', packageMap, files: {} };
    }
    this.store = store;
    this.saveState();
  }

  get packages(): PackageMap {
    return this.store.packageMap;
  }

  get files(): FileStateMap {
    return this.store.files;
  }

  // verify and update state based on files on disk
  getVerifiedStore(): Store {
    const store: Store = readJSONSync(this.configPath);
    for (const f in store.files) {
      const status = store.files[f];
      if (status.current && !existsSync(status.current)) {
        // if a ts file in state doesn't exist on disk, mark it as un-migrated
        store.files[f].current = null;
      }
      // update ts-ignore count
      store.files[f].errorCount = calculateTSIgnoreCount(f);
    }
    return store;
  }

  isStateExists(): boolean {
    if (!existsSync(DEFAULT_REHEARSAL_PATH)) {
      mkdirSync(DEFAULT_REHEARSAL_PATH);
    }
    return existsSync(this.configPath);
  }

  saveState(): void {
    writeJSONSync(this.configPath, this.store, { spaces: 2 });
  }

  addFilesToPackage(packageName: string, files: string[]): void {
    const fileMap: FileStateMap = {};
    // save files to package map for easier retrieve
    this.store.packageMap[packageName] = files;

    files.forEach(
      (f) =>
        (fileMap[f] = {
          origin: f.replace('.ts', '.js'),
          current: f,
          package: packageName,
          errorCount: calculateTSIgnoreCount(f as string),
        })
    );
    this.store.files = { ...this.store.files, ...fileMap };
    this.saveState();
  }

  getPackageMigrateProgress(packageName: string): PackageMigrateProgress {
    const fileList = this.store.packageMap[packageName];
    let migratedFileCount = 0;
    fileList.forEach((f) => {
      const fileState = this.store.files[f];
      if (fileState.current) {
        // if current is not null, then the migrated TS files should be there
        migratedFileCount++;
      }
    });
    return {
      migratedFileCount,
      totalFileCount: fileList.length,
      isCompleted: migratedFileCount === fileList.length,
    };
  }

  getPackageErrorCount(packageName: string): number {
    const fileList = this.store.packageMap[packageName];
    let errorCount = 0;
    fileList.forEach((f) => {
      const fileState = this.store.files[f];
      errorCount += fileState.errorCount;
    });
    return errorCount;
  }

  async addStateFileToGit(): Promise<void> {
    try {
      // check if git history exists
      await execa('git', ['status']);
      await execa('git', ['add', this.configPath]);
    } catch (e) {
      // no ops
    }
  }
}

export function calculateTSIgnoreCount(filePath: string): number {
  if (existsSync(filePath)) {
    const content = readFileSync(filePath, 'utf-8');
    return (content.match(REHEARSAL_TODO_REGEX) || []).length;
  } else {
    return 0;
  }
}
