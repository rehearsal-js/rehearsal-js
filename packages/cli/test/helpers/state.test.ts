import { resolve } from 'node:path';
import { existsSync, writeFileSync } from 'node:fs';
import { readJSONSync, writeJSONSync } from 'fs-extra/esm';
import { dirSync, setGracefulCleanup } from 'tmp';
import { beforeEach, describe, expect, test } from 'vitest';

import { State, calculateTSIgnoreCount, Store } from '../../src/helpers/state.js';

setGracefulCleanup();

function prepareTmpDir(): string {
  const { name: targetDir } = dirSync();
  return targetDir;
}

describe('state', () => {
  let basePath = '';

  beforeEach(() => {
    basePath = prepareTmpDir();
  });

  test('constructor should init state to disk', () => {
    const statePath = resolve(basePath, 'state.json');
    const { packages, files } = new State('foo', basePath, [], statePath);

    expect(packages).toEqual({});
    expect(files).toEqual({});
    expect(existsSync(statePath)).toBeTruthy();
    expect(readJSONSync(statePath)).matchSnapshot();
  });

  test('constructor should load existed state', () => {
    const statePath = resolve(basePath, 'state.json');
    const fooPath = resolve(basePath, 'foo');
    const existedStore: Store = {
      name: 'bar',
      packageMap: {
        bar: ['./foo'],
      },
      files: {},
    };
    existedStore.files['./foo'] = {
      origin: 'foo',
      current: 'foo',
      package: './bar',
      errorCount: 2,
    };

    // write existed state
    writeJSONSync(statePath, existedStore);
    // write foo file
    writeFileSync(fooPath, '');

    const { packages, files } = new State('foo', basePath, [], statePath);

    expect(packages['bar']).toEqual(['./foo']);
    expect(files['./foo'].package).toBe('./bar');
    expect(readJSONSync(statePath)).toMatchSnapshot();
  });

  test('getVerifiedStore', () => {
    const statePath = resolve(basePath, 'state.json');
    const existedStore: Store = {
      name: 'bar',
      packageMap: {
        './bar': ['./foo'],
      },
      files: {},
    };
    existedStore.files['./foo'] = {
      origin: 'foo',
      current: 'foo',
      package: './bar',
      errorCount: 2,
    };

    // write existed state
    writeJSONSync(statePath, existedStore);

    const { packages, files } = new State('foo', basePath, [], statePath);

    expect(packages['./bar']).toEqual(['./foo']);
    // there is no actual .foo existed on disk
    // stage would be updated when loading the old state file
    expect(files['./foo'].current).toBe(null);
    expect(readJSONSync(statePath)).toMatchSnapshot();
  });

  test('addFilesToPackages', () => {
    const statePath = resolve(basePath, 'state.json');
    const fooPath = resolve(basePath, 'foo');

    // write foo file
    writeFileSync(fooPath, '');

    const state = new State('bar', basePath, ['./bar'], statePath);

    state.addFilesToPackage('./bar', ['./foo']);

    expect(state.packages['./bar']).toEqual(['./foo']);
    expect(state.files['./foo'].package).toBe('./bar');
    expect(readJSONSync(statePath)).toMatchSnapshot();
  });

  test('calculateTSIgnoreCount', () => {
    const foo = '@rehearsal TODO foo bar';
    const fooPath = resolve(basePath, 'foo');

    const bar = '@rehearsal TODO foo\n@rehearsal TODO bar';
    const barPath = resolve(basePath, 'bar');

    writeFileSync(fooPath, foo, 'utf-8');
    writeFileSync(barPath, bar, 'utf-8');

    expect(calculateTSIgnoreCount(fooPath)).toBe(1);
    expect(calculateTSIgnoreCount(barPath)).toBe(2);
  });

  test('getPackageMigrateProgress', () => {
    const statePath = resolve(basePath, 'state.json');
    const fooPath = resolve(basePath, 'foo.ts');
    const store: Store = {
      name: 'bar',
      packageMap: {
        bar: ['./foo.ts', './bar.ts'],
      },
      files: {},
    };
    store.files['./foo.ts'] = {
      origin: './foo.ts',
      current: './foo.ts',
      package: 'bar',
      errorCount: 2,
    };
    store.files['./bar.ts'] = {
      origin: './bar.ts',
      current: null,
      package: 'bar',
      errorCount: 2,
    };

    writeJSONSync(statePath, store);
    // only have fooPath on disk, so state should know only foo is migrated
    writeFileSync(fooPath, '', 'utf-8');

    let state = new State('bar', basePath, ['sample-package'], statePath);
    state.addFilesToPackage('sample-package', ['./foo.ts', './bar.ts']);
    // here create the state again, to simulate loading previous state
    // the getVerifiedStore would be triggered here to update files' state
    state = new State('bar', basePath, ['sample-package'], statePath);

    const expected = {
      migratedFileCount: 1,
      totalFileCount: 2,
      isCompleted: false,
    };

    expect(state.getPackageMigrateProgress('sample-package')).toStrictEqual(expected);
    expect(readJSONSync(statePath)).toMatchSnapshot();
  });

  test('getPackageErrorCount', () => {
    const statePath = resolve(basePath, 'state.json');
    const foo = '@rehearsal TODO foo bar';
    const fooPath = resolve(basePath, 'foo');
    const bar = '@rehearsal TODO foo\n@rehearsal TODO bar';
    const barPath = resolve(basePath, 'bar');
    const state = new State('bar', basePath, ['sample-package'], statePath);

    writeFileSync(fooPath, foo, 'utf-8');
    writeFileSync(barPath, bar, 'utf-8');

    state.addFilesToPackage('sample-package', [fooPath, barPath]);

    expect(state.getPackageErrorCount('sample-package')).toStrictEqual(3);
  });

  test('does not contain absolute paths in state file', () => {
    const statePath = resolve(basePath, 'state.json');
    const fooPath = resolve(basePath, 'foo');
    const barPath = resolve(basePath, 'bar');
    const packagePath = resolve(basePath, 'sample-package');

    const state = new State('bar', basePath, [packagePath], statePath);
    // check state file after init
    expect(readJSONSync(statePath)).toMatchSnapshot();

    state.addFilesToPackage(packagePath, [fooPath, barPath]);
    // check state file again after addFilesToPackage
    expect(readJSONSync(statePath)).toMatchSnapshot();
  });
});
