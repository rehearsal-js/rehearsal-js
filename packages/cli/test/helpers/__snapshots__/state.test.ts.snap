// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`state > addFilesToPackages 1`] = `
{
  "files": {
    "./foo": {
      "current": "./foo",
      "errorCount": 0,
      "origin": "./foo",
      "package": "./bar",
    },
  },
  "name": "bar",
  "packageMap": {
    "./bar": [
      "./foo",
    ],
  },
}
`;

exports[`state > constructor should init state to disk 1`] = `
{
  "files": {},
  "name": "foo",
  "packageMap": {},
}
`;

exports[`state > constructor should load existed state 1`] = `
{
  "files": {
    "./foo": {
      "current": "foo",
      "errorCount": 0,
      "origin": "foo",
      "package": "./bar",
    },
  },
  "name": "bar",
  "packageMap": {
    "bar": [
      "./foo",
    ],
  },
}
`;

exports[`state > does not contain absolute paths in state file 1`] = `
{
  "files": {},
  "name": "bar",
  "packageMap": {
    "./sample-package": [],
  },
}
`;

exports[`state > does not contain absolute paths in state file 2`] = `
{
  "files": {
    "./bar": {
      "current": null,
      "errorCount": 0,
      "origin": "./bar",
      "package": "./sample-package",
    },
    "./foo": {
      "current": null,
      "errorCount": 0,
      "origin": "./foo",
      "package": "./sample-package",
    },
  },
  "name": "bar",
  "packageMap": {
    "./sample-package": [
      "./foo",
      "./bar",
    ],
  },
}
`;

exports[`state > getPackageMigrateProgress 1`] = `
{
  "files": {
    "./bar.ts": {
      "current": null,
      "errorCount": 0,
      "origin": "./bar.ts",
      "package": "sample-package",
    },
    "./foo.ts": {
      "current": "./foo.ts",
      "errorCount": 0,
      "origin": "./foo.ts",
      "package": "sample-package",
    },
  },
  "name": "bar",
  "packageMap": {
    "bar": [
      "./foo.ts",
      "./bar.ts",
    ],
    "sample-package": [
      "./foo.ts",
      "./bar.ts",
    ],
  },
}
`;

exports[`state > getVerifiedStore 1`] = `
{
  "files": {
    "./foo": {
      "current": null,
      "errorCount": 2,
      "origin": "foo",
      "package": "./bar",
    },
  },
  "name": "bar",
  "packageMap": {
    "./bar": [
      "./foo",
    ],
  },
}
`;
